dohproxy/server_protocol.py

```diff
@@ -42,10 +43,34 @@ def __init__(self, upstream_resolver, upstream_port, logger=None):
         self.logger = logger
         self.transport = None
 
-    async def query(self, dnsq, clientip, timeout=DEFAULT_TIMEOUT):
-        dnsr = await self.query_udp(dnsq, clientip, timeout=timeout)
+    async def query(self, dnsq, clientip, timeout=DEFAULT_TIMEOUT,
+                    ecs=False):
+        original_ecs_option = None
+        for option in dnsq.options:
+            if isinstance(option, dns.edns.ECSOption):
+                original_ecs_option = option
+                break
+
+        # (Potentially) modified copy of dnsq
+        dnsq_mod = dns.message.from_wire(dnsq.to_wire())
+        if ecs:
+            utils.set_dns_ecs(dnsq_mod, clientip)
+
+        dnsr = await self.query_udp(dnsq_mod, clientip, timeout=timeout)
         if dnsr is None or (dnsr.flags & dns.flags.TC):
-            dnsr = await self.query_tcp(dnsq, clientip, timeout=timeout)
+            dnsr = await self.query_tcp(dnsq_mod, clientip, timeout=timeout)
+
+        if dnsr is not None and ecs:
+            options = []
+            for option in dnsr.options:
+                if isinstance(option, dns.edns.ECSOption):
+                    if original_ecs_option is not None:
+                        options.append(original_ecs_option)
+                else:
+                    options.append(option)
+            dnsr.edns = dnsq.edns
+            dnsr.options = options
```

Comment:
same here, we only need to "reset" EDNS *if* we modified it in the first place.

Reply:
Agreed, it's now part of "if dnsr is not None and we_set_ecs".